src/abstract_domains.rs

```diff
@@ -258,6 +263,12 @@ impl AbstractDomain for ExpressionDomain {
 
     /// Returns an expression that is "self == other".
     fn equals(&self, other: &Self) -> Self {
+        if self == other {
+            if self.is_bottom() || self.is_top() {
+                return self.clone();
+            };
+            return ExpressionDomain::CompileTimeConstant(ConstantValue::True);
```

Comment:
Not sure I get this, why is it ConstantValue::True in this case?

Reply:
It's a good question and I'm not totally sure of the answer. Intuitively, for any (functional) expression x, we should have x == x. Except of course if x has the runtime value of NaN. So why the special behavior for Top and Bottom? Using Bottom should always be an error, so returning Bottom seems more appropriate. Top can be a generalization of many different expressions, so self == Top and other == Top does not seem the same as x == x.

Upon further reflection this is a deeply unsatisfying answer. At the time it seemed like a good idea for equals to be in line with subset, but the operations are actually quite different. Since I don't need this change to make the fixed point terminate, I'm just going to remove it.

Thanks for noticing and speaking up about it.